OpenRA.Mods.Common/Widgets/Logic/Settings/HotkeysSettingsLogic.cs

```diff
@@ -144,21 +165,67 @@ Action ResetPanel(Widget panel)
 			};
 		}
 
+		void InitHotkeyList()
+		{
+			hotkeyList.RemoveChildren();
+			selectedHotkeyDefinition = null;
+
+			foreach (var hg in hotkeyGroups)
+			{
+				var typesInGroup = hg.Value;
+				var keysInGroup = modData.Hotkeys.Definitions
+					.Where(hd => IsHotkeyVisibleInFilter(hd) && hd.Types.Overlaps(typesInGroup));
+
+				if (!keysInGroup.Any())
+					continue;
+
+				var header = headerTemplate.Clone();
+				header.Get<LabelWidget>("LABEL").GetText = () => hg.Key;
+				hotkeyList.AddChild(header);
+
+				var added = new HashSet<HotkeyDefinition>();
+
+				foreach (var type in typesInGroup)
+				{
+					foreach (var hd in keysInGroup.Where(k => k.Types.Contains(type)))
+					{
+						if (added.Add(hd))
+						{
+							if (selectedHotkeyDefinition == null)
+								selectedHotkeyDefinition = hd;
+
+							BindHotkeyPref(hd, template);
+						}
+					}
+				}
+			}
+
+			emptyListMessage.Visible = selectedHotkeyDefinition == null;
+			remapDialog.Visible = selectedHotkeyDefinition != null;
+
+			hotkeyList.ScrollToTop();
+		}
+
 		void InitHotkeyRemapDialog(Widget panel)
 		{
-			var label = new CachedTransform<HotkeyDefinition, string>(hd => hd.Description + ":");
-			panel.Get<LabelWidget>("HOTKEY_LABEL").GetText = () => label.Update(selectedHotkeyDefinition);
+			var label = panel.Get<LabelWidget>("HOTKEY_LABEL");
+			var labelText = new CachedTransform<HotkeyDefinition, string>(hd => hd?.Description + ":");
+			label.IsVisible = () => selectedHotkeyDefinition != null;
+			label.GetText = () => labelText.Update(selectedHotkeyDefinition);
 
 			var duplicateNotice = panel.Get<LabelWidget>("DUPLICATE_NOTICE");
 			duplicateNotice.TextColor = ChromeMetrics.Get<Color>("NoticeErrorColor");
 			duplicateNotice.IsVisible = () => !isHotkeyValid;
-			var duplicateNoticeText = new CachedTransform<HotkeyDefinition, string>(hd => hd != null ? duplicateNotice.Text.F(hd.Description) : duplicateNotice.Text);
+			var duplicateNoticeText = new CachedTransform<HotkeyDefinition, string>(hd =>
+				hd != null ?
+				duplicateNotice.Text.F(hd.Description, hd.Contexts.First(c => selectedHotkeyDefinition.Contexts.Contains(c))) :
+				duplicateNotice.Text);
```

Comment:
The duplicate notice has `{0}` and `{1}` in it, so does it even make sense to display that here without `.F`? Also, if `hd` is `null` in this context, then `duplicateHotkeyDefinition` is also `null`, correct? In that case this text is not displayed anyway, so I'd suggest putting an empty string.